core/src/main/java/org/bitcoinj/base/Bech32.java

```diff
@@ -51,15 +54,71 @@ public class Bech32 {
 
     public enum Encoding { BECH32, BECH32M }
 
-    public static class Bech32Data {
+    public static class Bech32Bytes extends ByteArray {
+        /**
+         * Wrapper for a {@code byte[]} array.
+         *
+         * @param bytes bytes to be copied
+         */
+        protected Bech32Bytes(byte[] bytes) {
+            super(bytes);
+        }
+
+        private Bech32Bytes(byte first, byte[] rest) {
```

Comment:
I think the plain Bech32 spec (just the ["Bech" chapter](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32) of BIP173), and it doesn't have this "first byte + rest" concept. It's just hrp, separator (irrelevant) and data.

So I think a Bech32Bytes class should probably not have this, at least not without a good reason. Maybe a more specialized "SegwitBech32Bytes" class on top would make sense?

Reply:
I just read through the Nostr ["bech32-encoded entities" spec](https://github.com/nostr-protocol/nips/blob/master/19.md).

It seems to not have any mention of "first byte" or "version byte". It also always uses Bech32 (_not_ "m") encoding. Segwit ties the encoding strictly to the version byte: 0 uses Bech32, >0 uses Bech32m.

Reply:
> … the ["Bech" chapter](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32) of BIP173, doesn't have this "first byte + rest" concept.

True, but this is really just an alternate representation of a `byte[]`. So it doesn't really violate in "layering" principles in the spec.

> without a good reason

This is a convenience constructor (that simplifies the code and avoids a byte[] copy) for the Segwit implementation but without using any Segwit types.

Note that it is `private`.

> Maybe a more specialized "SegwitBech32Bytes" or "VersionedBech32Bytes" class on top would make sense?

Personally, I think we have enough classes in this implementation already and feel the convenience constructor is the simpler solution.

Maybe trying to eliminate the copy is unnecessary optimization (though there are applications that might deal with thousands or millions of `Address` and need to serialize them.)

I originally coded it without this method and added it later, so if you insist we could break it into multiple commits and maybe when you see it you'll agree with me. I also consider having the first parameter be a `short` or an `int` instead of a `byte`, but that seems to violate "layering" a little more.

Reply:
> the Nostr ["bech32-encoded entities" spec](https://github.com/nostr-protocol/nips/blob/master/19.md).

> … seems to not have any mention of "first byte" or "version byte".

That is true. This constructor was added for the refactored `SegwitAddress` implementation.

Reply:
Ok, I can live with this one. Bech32 is close enough to segwit that some of its artifacts can spill over.

But there is also its cousin in ByteArray, see my other comment.

Reply:
Fixed in new commits.